Dynamic inline workflow def suboworfklow

c4lm opened the pull request:
Pull Request type
----
- [x] Bugfix

Changes in this PR
----

_Describe the new behavior from this PR, and why it's needed_
Currently unsupported.
Issue [#1982 of parent OSS repo](https://github.com/Netflix/conductor/issues/1982)
This PR will allow setting subWorkflowParams like this:
```
"subWorkflowParam": {
        "workflowDefinition": "${workflow.input.tst}"
      }
```
so that parameter resolution will happen, and workflowDefinition, if available, will take priority over name.

Alternatives considered
----

The alternative was to keep everything in place, but alter SubWorkflowTaskMapper and SubWorkflow so that rname/version/definition resolution actually comes from task input instead of `subWorkflowParams`.

c4lm committed:
Add support for subworkflow tasks which have SubworkflowParam's workflowDefinition coming from input (inline)

c4lm committed:
Add support for subworkflow tasks which have SubworkflowParam's workflowDefinition coming from input (inline)
spotlessApply

c4lm committed:
Add support for subworkflow tasks which have SubworkflowParam's workflowDefinition coming from input (inline)
change in serializtion and validations

c4lm committed:
Add support for subworkflow tasks which have SubworkflowParam's workflowDefinition coming from input (inline)
fix constraint test, spotlessApply misc

c4lm commented on common/src/main/java/com/netflix/conductor/common/utils/TaskUtils.java:
```diff
@@ -28,4 +35,8 @@ public static String removeIterationFromTaskRefName(String referenceTaskName) {
         return tokens.length > 0 ? tokens[0] : referenceTaskName;
     }
+
+    public static WorkflowDef convertToWorkflowDef(Object workflowDef) {
```
can't have static objectMapper without screwing up generation of protomapper, so it is here

c4lm commented on common/src/main/java/com/netflix/conductor/common/metadata/workflow/SubWorkflowParams.java:
```diff
@@ -12,24 +12,18 @@
 public class SubWorkflowParams {
 
     @ProtoField(id = 1)
-    @NotNull(message = "SubWorkflowParams name cannot be null")
```
Validation moves to  WorkflwoTask::setSubworkflowParam

coderabhigupta commented on core/src/main/java/com/netflix/conductor/core/execution/tasks/SubWorkflow.java:
```diff
@@ -56,6 +58,16 @@ public void start(WorkflowModel workflow, TaskModel task, WorkflowExecutor workf
+                    Optional.ofNullable(input.get("subWorkflowName"))
+                            .map(Object::toString)
+                            .orElse(null);
+            version = (Integer) input.get("subWorkflowVersion");
```
Should we default to latest version in case it is missing from the input?

c4lm replied:
this is done in taskmapper

coderabhigupta commented on common/src/main/java/com/netflix/conductor/common/metadata/workflow/SubWorkflowParams.java:
```diff
@@ -100,31 +94,28 @@ public Object getWorkflowDefinition() {
+        } else {
             throw new IllegalArgumentException(
-                    "workflowDefinition must be either null or WorkflowDef");
+                    "workflowDefinition must be either null, or WorkflowDef, or a valid DSL string");
```
Shouldn't you add  `or a LinkedHashMap` in this exception?

c4lm replied:
This exception will propagate to UI on workflow save. If I add LinkedHashMap It will only confuse users and will not tell them anything useful, because for the users workflowDefinition **cannot** be just a LinkedHashMap.

c4lm committed:
Add support for subworkflow tasks which have SubworkflowParam's workflowDefinition coming from input (inline)
change taskmapper logic for workflowDefinition

c4lm committed:
Add support for subworkflow tasks which have SubworkflowParam's workflowDefinition coming from input (inline)
dummy commit to trigger github sync

c4lm commented on core/src/main/java/com/netflix/conductor/core/execution/mapper/SubWorkflowTaskMapper.java:
```diff
@@ -116,12 +122,17 @@ private Map<String, Object> getSubWorkflowInputParameters(
-        Object subWorkflowDefinition = subWorkflowParams.getWorkflowDefinition();
-        if (subWorkflowDefinition != null) {
-            params.put("workflowDefinition", subWorkflowDefinition);
+        if (subWorkflowParams.getWorkflowDefinition() == null
+                || (subWorkflowParams.getWorkflowDefinition() != null
+                        && subWorkflowParams.getWorkflowDefinition() instanceof String)) {
+            params.put("workflowDefinition", subWorkflowParams.getWorkflowDefinition());
+            params = parametersUtils.getTaskInputV2(params, workflowModel, null, null);
+        } else {
+            // not DSL but actual workflowDef - do not resolve params inside subworkflow definition
+            Object subWorkflowDefinition = subWorkflowParams.getWorkflowDefinition();
+            if (subWorkflowDefinition != null) {
+                params.put("workflowDefinition", subWorkflowDefinition);
+            }
         }
```
Seems like github is experiencing sync issues, I updated the fork branch but the commits do not show up. I'll wait a bit and send one more dummy commit later to see if it'll trigger then

c4lm replied:
^ both are wrong
the call to parameter utils has to be done exactly once

c4lm committed:
Add support for subworkflow tasks which have SubworkflowParam's workflowDefinition coming from input (inline)
spotless

c4lm committed:
Add support for subworkflow tasks which have SubworkflowParam's workflowDefinition coming from input (inline)
spotless & return name from definition when possible, like before for compatibility

c4lm committed:
Add support for subworkflow tasks which have SubworkflowParam's workflowDefinition coming from input (inline)
fix parameter resolution

c4lm committed:
Add support for subworkflow tasks which have SubworkflowParam's workflowDefinition coming from input (inline)
spotless

c4lm committed:
Add support for subworkflow tasks which have SubworkflowParam's workflowDefinition coming from input (inline)
also return version with the same logic

c4lm committed:
Add support for subworkflow tasks which have SubworkflowParam's workflowDefinition coming from input (inline)

fix taskmapper again

c4lm commented on common/src/main/java/com/netflix/conductor/common/metadata/workflow/SubWorkflowParams.java:
```diff
@@ -100,31 +96,28 @@ public Object getWorkflowDefinition() {
+        } else if (workflowDef instanceof WorkflowDef) {
+            this.workflowDefinition = workflowDef;
+        } else if (workflowDef instanceof String) {
+            if (!(((String) workflowDef).startsWith("${"))
```
Thought to make it a regex but after thinking more about it - this is fine, there is some room in what is allowed as a valid DSL string for parameter utils
